lib/explorer/shared.ex

```diff
@@ -277,6 +304,31 @@ defmodule Explorer.Shared do
     dtype_from_list!(items, leaf_dtype(type))
   end
 
+  defp new_type_matches?(type, new_type)
+
+  defp new_type_matches?(nil, _new_trpe), do: true
+
+  defp new_type_matches?({:struct, types}, {:struct, new_types}) do
+    types
+    |> Map.keys()
+    |> Enum.all?(fn key ->
+      type = Map.get(types, key)
+
+      case Map.fetch(new_types, key) do
+        {:ok, new_type} -> new_type_matches?(type, new_type)
+        :error -> false
+      end
+    end)
+  end
+
+  defp new_type_matches?(type, new_type) do
+    if leaf_dtype(new_type) == :numeric and leaf_dtype(type) in [:integer, {:f, 32}, {:f, 64}] do
+      true
+    else
+      type == new_type
+    end
```

Comment:
Define a `  defp new_type_matches?(type, type), do: true` at the top and this code can become:

```suggestion
    leaf_dtype(new_type) == :numeric and leaf_dtype(type) in [:integer, {:f, 32}, {:f, 64}]
```

Reply:
Thanks, looks a lot cleaner now!

Added in [53e3e87](https://github.com/elixir-explorer/explorer/pull/756/commits/53e3e8727369475c78c9242267dcc1867c0192f9)